source/mir/random/tinflex/internal/types.d

```diff
@@ -0,0 +1,318 @@
+module mir.random.tinflex.internal.types;
+
+import std.traits: ReturnType, isFloatingPoint;
+
+protected:
+
+/**
+Major data unit of the Tinflex algorithm.
+It is used to store
+- (cached) values of the transformation (and its derivatives)
+- area below the hat and squeeze function
+- linked-list like reference to the right part of the interval (there will always
+be exactly one interval with right = 0)
+*/
+struct IntervalPoint(S)
+    if (isFloatingPoint!S)
+{
+    import mir.random.tinflex.internal.linearfun : LinearFun;
+
+    /// left position of the interval
+    immutable S x;
+
+    /// T_c family of the interval
+    immutable S c;
+
+    /// transformed values
+    immutable S tx, t1x, t2x;
+
+    this (in S tx, in S t1x, in S t2x, in S x, in S c)
```

Comment:
> Constructor does not required (it has just reordering)

My motivation was that the invariant check (against NaNs) is called after the constructor.